### PR TITLE
creating migration for invoice,invoice_products and invoice_payment a…

### PR DESCRIPTION
**Pull request by HamidNiakan**
…nd then creating class as InvoiceManager for CRUD

### CODE REVIEW

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+         Schema::create('invoice_products', function (Blueprint $table) {
+             $floatScale = $this->getFloatScale();
+             $table->id();
+             $table->string('title', 255);
+             $table->foreignId('invoice_id');
+             $table->foreignId('currency_id');
+             $table->timestamps();
+             $table->decimal('price', 10 + $floatScale, $floatScale);
+             $table->decimal('discount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('total_amount', 10 + $floatScale, $floatScale)->nullable();
```
Comment (yeganemehr) [inline database/migrations/2023_01_18_000002_create_invoices_products_table.php:22]: calculate this in your model and don't save in it in the table.

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+             $table->integer('count');
+             $table->json('meta')->nullable();
+             $table->json('distribution_plan')->nullable();
+             $table->json('distribution')->nullable();
+             $table->longText('description')->nullable();
```
Comment (yeganemehr) [inline database/migrations/2023_01_18_000002_create_invoices_products_table.php:27]: use text. descriptions in invoice items are not that long

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+             $table->foreign('Invoice_id')
```
Comment (yeganemehr) [inline database/migrations/2023_01_18_000002_create_invoices_products_table.php:29]: merge these `foreign()`s with above `foreignId()`s

File: config/invoice.php [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+return [
+    // Define your user model class for connect tickets to users.
```
Comment (yeganemehr) [inline config/invoice.php:4]: tickets?

File: database/factories/InvoiceFactory.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    public function withEUR()
+    {
+        return $this->withCurrency(Currency::factory()
+                                           ->create()
+                                           ->asEUR());
+    }
+
+    public function paid()
```
Comment (yeganemehr) [inline database/factories/InvoiceFactory.php:116]: when it's paid, it's must has paid_time

File: database/migrations/2023_01_18_000003_create_invoices_payments_table.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+     {
+         Schema::create('invoice_payments', function (Blueprint $table) {
+             $floatScale = $this->getFloatScale();
+             $table->id();
+             $table->foreignId('invoice_id');
+             $table->foreignId('transaction_id')->nullable();
+             $table->timestamps();
+             $table->string('method');
+             $table->decimal('amount', 10 + $floatScale, $floatScale);
+             $table->string('status');
```
Comment (yeganemehr) [inline database/migrations/2023_01_18_000003_create_invoices_payments_table.php:21]: use enums

File: database/migrations/2023_01_18_000003_create_invoices_payments_table.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ use Illuminate\Database\Migrations\Migration;
+ use Illuminate\Database\Schema\Blueprint;
+ use Illuminate\Support\Facades\Schema;
+
+ return new class() extends Migration {
+     use ModelHelpers;
+
+     public function up(): void
+     {
+         Schema::create('invoice_payments', function (Blueprint $table) {
```
Comment (yeganemehr) [inline database/migrations/2023_01_18_000003_create_invoices_payments_table.php:13]:
```suggestion
         Schema::create('invoices_payments', function (Blueprint $table) {
```

File: composer.json [hunk 8]
```diff
@@ -13,7 +14,9 @@
 	},
 	"require": {
 		"php": "^8.1",
-		"dnj/laravel-account": "dev-master"
+		"dnj/laravel-account": "dev-master",
+		"dnj/laravel-currency": "^1.0.1",
+		"laravel/legacy-factories": "^1.3"
```
Comment (yeganemehr) [inline composer.json:19]: why did you need legacy-factories?

File: database/factories/PaymentFactory.php [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+{
+    protected $model = Payment::class;
+
+    public function definition()
+    {
+        // TODO: Implement definition() method.
+        return [
+            'invoice_id' => Invoice::factory(),
+            'transaction_id' => null,
+            'method' => fake()->sentence(3),
```
Comment (yeganemehr) [inline database/factories/PaymentFactory.php:22]: it's should be class name which implemented `Contracts\IPaymentMethod` interface

File: database/factories/ProductFactory.php [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    public function withCurrency(Currency|CurrencyFactory $currency)
+    {
+        return $this->state(fn () => [
+            'currency_id' => $currency,
+        ]);
+    }
+
+    public function withTile(string $title)
```
Comment (yeganemehr) [inline database/factories/ProductFactory.php:40]:
```suggestion
    public function withTitle(string $title)
```

File: database/factories/ProductFactory.php [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    public function withCount(int $count)
+    {
+        return $this->state(fn () => [
+            'count' => $count,
+        ]);
+    }
+
+    public function withInvoice(int $invoiceId)
```
Comment (yeganemehr) [inline database/factories/ProductFactory.php:82]: Let's accept Invoice|InvoiceFactory instance instead of just id.

File: src/InvoiceManager.php [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+    public function approvePayment(int $paymentId): IPayment
+    {
+        // TODO: Implement approvePayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        if ($payment->getAmount() > $payment->invoice->getAmount()) {
+            throw new AmountInvoiceMismatchException();
```
Comment (yeganemehr) [inline src/InvoiceManager.php:289]: You check amount of payment after you inserted in the database?

File: src/InvoiceManager.php [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+        }
+        if ($payment->getAmount() == $payment->invoice->getAmount()) {
+            $payment->update([
+                                 'status' => PaymentStatus::APPROVED,
```
Comment (yeganemehr) [inline src/InvoiceManager.php:293]: you just approve payment just when client pay invoice in one part?

File: src/InvoiceManager.php [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+
+    public function rejectPayment(int $paymentId): IPayment
+    {
+        // TODO: Implement rejectPayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment->update([
```
Comment (yeganemehr) [inline src/InvoiceManager.php:312]: What if payment already approved? no need to check status of payment?

File: src/InvoiceManager.php [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+
+        return $payment;
+    }
+
+    public function getInvoiceById(int $invoiceId): Invoice
+    {
+        // TODO: Implement getById() method.
+        return Invoice::query()
+                      ->findOrFail($invoiceId)
+                      ->load('products');
```
Comment (yeganemehr) [inline src/InvoiceManager.php:324]: if you want to [eager loading](https://laravel.com/docs/9.x/eloquent-relationships#eager-loading), put it in the model.

File: src/InvoiceManager.php [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+    public function addPaymentToInvoice(int $invoiceId, string $type, INumber $amount, PaymentStatus $status, ?array $meta): IPayment
+    {
+        // TODO: Implement addPaymentToInvoice() method.
+        $invoice = $this->getInvoiceById($invoiceId);
+        if (InvoiceStatus::PAID == $invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment = $invoice->payments()
+                           ->create([
+                                        'method' => $type,
```
Comment (yeganemehr) [inline src/InvoiceManager.php:271]: Did you check type is valid class name that is implemented `IPaymentMethod` interface?

File: src/InvoiceManager.php [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+
+    public function create(int $userId, int $currencyId, array $products, array $localizedDetails, ?array $meta): IInvoice
+    {
+        // TODO: Implement create() method.
+
+        $invoice = new Invoice();
+        $invoice->user_id = $userId;
+        $invoice->currency_id = $currencyId;
+        $invoice->meta = $meta;
+        if (isset($localizedDetails['title'])) {
```
Comment (yeganemehr) [inline src/InvoiceManager.php:35]: title is not optional

File: src/InvoiceManager.php [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+            $invoice->title = $localizedDetails['title'];
+        }
+        $invoice->status = InvoiceStatus::UNPAID->value;
```
Comment (yeganemehr) [inline src/InvoiceManager.php:38]:
```suggestion
        $invoice->status = InvoiceStatus::UNPAID;
```

File: src/InvoiceManager.php [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+        $invoice->save();
+        $invoice->products()
+                ->createMany($this->buliderInvoiceProducts($products));
+        $invoice->update([
```
Comment (yeganemehr) [inline src/InvoiceManager.php:42]: update after creation? :exploding_head:

File: src/InvoiceManager.php [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+        }
+        if ($invoice->products->isNotEmpty()) {
+            $invoice->products()
+                    ->delete();
+        }
+        if ($invoice->payments->isNotEmpty()) {
+            $invoice->payments()
+                    ->delete();
+        }
+        $invoice->delete();
```
Comment (yeganemehr) [inline src/InvoiceManager.php:66]: just make relations cascade on delete and delete invoice. no need to delete payments and products separately.

File: src/InvoiceManager.php [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+
+    public function addProductToInvoice(int $invoiceId, array $product): IProduct
+    {
+        // TODO: Implement addProductToInvoice() method.
+
+        $invoice = $this->getInvoiceById($invoiceId);
+
+        return DB::transaction(function () use ($invoice, $product) {
+            $record = new Product();
+            if (isset($product['price'])) {
```
Comment (yeganemehr) [inline src/InvoiceManager.php:112]: can we add a product to the invoice without price or title or count or ...?

File: src/ModelHelpers.php [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+        if ($userModel) {
+            $userTable = (new $userModel())->getTable();
+        }
+
+        return $userTable;
+    }
+
+    protected function getCurrencyTable(): string
+    {
+        return (new Currency())->getTable();
```
Comment (yeganemehr) [inline src/ModelHelpers.php:28]: You can use Currency model statically

**Review by yeganemehr — CHANGES_REQUESTED** [review]

**Comment by yeganemehr** [comment]
There is MUCH to change
I just got tired of this
waste of time

File: database/migrations/2023_01_18_000001_create_invoices_table.php [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+             $table->id();
+             $table->string('title', 255);
+             $table->foreignId('user_id');
+             $table->foreignId('currency_id');
+             $table->timestamps();
+             $table->decimal('amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('paid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('unpaid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->json('meta')->nullable();
+             $table->dateTime('paid_time')->nullable();
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000001_create_invoices_table.php:24]: done

File: database/migrations/2023_01_18_000001_create_invoices_table.php [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+         Schema::create('invoices', function (Blueprint $table) {
+             $floatScale = $this->getFloatScale();
+             $table->id();
+             $table->string('title', 255);
+             $table->foreignId('user_id');
+             $table->foreignId('currency_id');
+             $table->timestamps();
+             $table->decimal('amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('paid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('unpaid_amount', 10 + $floatScale, $floatScale)->nullable();
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000001_create_invoices_table.php:22]: done

File: src/Models/Invoice.php [hunk 25]
```diff
@@ -150,17 +155,22 @@ public function recalculateTotalAmount(): void
          * @var IExchangeManager
          */
         $exchange = app(IExchangeManager::class);
-
         $result = Number::fromInput(0);
         foreach ($this->getProducts() as $product) {
             $amountInSameCurrency = $exchange->convert($product->getTotalAmount(), $this->getCurrencyId(), $product->getCurrencyId(), true);
             $result = $result->add($amountInSameCurrency);
         }
+        dd($result);
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply src/Models/Invoice.php:163]: done

File: tests/Unit/InvoiceManagerTest.php [hunk 26]
```diff
@@ -0,0 +1,429 @@
...
+        $data = [
+            'title' => 'update invoice one',
+            'user_id' => $user->id,
+            'meta' => [
+                'key1' => 'value',
+            ],
+            'currencyId' => $EUR->getID(),
+            'products' => [
+                [
+                    'id' => 1,
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply tests/Unit/InvoiceManagerTest.php:79]: done

File: tests/Unit/InvoiceManagerTest.php [hunk 27]
```diff
@@ -0,0 +1,429 @@
...
+
+    /**
+     * Testing create invoice.
+     */
+    public function testCreateInvoice(): void
+    {
+        $now = time();
+        $user = User::factory()
+                    ->create();
+        $USD = $this->createUSD();
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply tests/Unit/InvoiceManagerTest.php:31]: done

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+ use Illuminate\Database\Migrations\Migration;
+ use Illuminate\Database\Schema\Blueprint;
+ use Illuminate\Support\Facades\Schema;
+
+ return new class() extends Migration {
+     use ModelHelpers;
+
+     public function up(): void
+     {
+         Schema::create('invoice_products', function (Blueprint $table) {
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000002_create_invoices_products_table.php:13]: done

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php (hunk 2 above)
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000002_create_invoices_products_table.php:27]: done

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php (hunk 3 above)
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000002_create_invoices_products_table.php:29]: done

File: config/invoice.php (hunk 4 above)
Comment (HamidNiakan) [reply config/invoice.php:4]: done

File: database/migrations/2023_01_18_000003_create_invoices_payments_table.php (hunk 7 above)
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000003_create_invoices_payments_table.php:13]: done

File: composer.json (hunk 8 above)
Comment (HamidNiakan) [reply composer.json:19]: This was for the previous changes I made to the User Factory
done

File: database/factories/ProductFactory.php (hunk 10 above)
Comment (HamidNiakan) [reply database/factories/ProductFactory.php:40]: done

File: database/factories/ProductFactory.php (hunk 11 above)
Comment (HamidNiakan) [reply database/factories/ProductFactory.php:82]: done

File: src/InvoiceManager.php (hunk 15 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php:324]: done

File: src/InvoiceManager.php (hunk 17 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php:35]: done

File: src/InvoiceManager.php (hunk 18 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php:38]: done

File: src/InvoiceManager.php (hunk 21 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php:112]: done

File: database/migrations/2023_01_18_000003_create_invoices_payments_table.php (hunk 6 above)
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000003_create_invoices_payments_table.php:21]: done

File: database/migrations/2023_01_18_000001_create_invoices_table.php [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+             $table->json('meta')->nullable();
+             $table->dateTime('paid_time')->nullable();
+             $table->string('status')->nullable();
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000001_create_invoices_table.php:25]: done

File: src/InvoiceManager.php (hunk 20 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php:66]: done

File: database/migrations/2023_01_18_000001_create_invoices_table.php [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+
+             $table->foreign('currency_id')
```
[the remark this answers is not in the document]
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000001_create_invoices_table.php:27]: done

File: database/factories/InvoiceFactory.php (hunk 5 above)
Comment (HamidNiakan) [reply database/factories/InvoiceFactory.php:116]: done

File: database/migrations/2023_01_18_000002_create_invoices_products_table.php (hunk 1 above)
Comment (HamidNiakan) [reply database/migrations/2023_01_18_000002_create_invoices_products_table.php:22]: done

File: src/InvoiceManager.php (hunk 19 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php]: done

File: src/InvoiceManager.php (hunk 12 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php]: done

File: src/InvoiceManager.php (hunk 13 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php]: done

File: src/InvoiceManager.php (hunk 14 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php]: done

File: src/ModelHelpers.php (hunk 22 above)
Comment (HamidNiakan) [reply src/ModelHelpers.php]: done

File: database/factories/PaymentFactory.php (hunk 9 above)
Comment (HamidNiakan) [reply database/factories/PaymentFactory.php]: I can't implement it

File: src/InvoiceManager.php (hunk 16 above)
Comment (HamidNiakan) [reply src/InvoiceManager.php]: I can't implement it
